bin/weldr/src/convert.rs

```diff
@@ -107,154 +103,261 @@ impl ConvertCommand {
-        let line_indices_bytes: &[u8] = unsafe { as_u8_slice(line_indices) };
-        let triangle_indices = &geometry_cache.triangle_indices[..];
-        let triangle_indices_bytes: &[u8] = unsafe { as_u8_slice(triangle_indices) };
-        self.write_impl(
-            &[vertices_bytes, line_indices_bytes, triangle_indices_bytes],
-            "glbuf",
-        )?;
-
-        // Prepare geometry buffers for glTF writing
-        let mut offset: usize = 0;
-        let vertex_buffer = GeometryBuffer {
-            size: vertices_bytes.len(),
-            offset,
-            stride: 12,
-            component_type: gltf::ComponentType::Float,
-            attribute_type: gltf::AttributeType::Vec3,
+    fn add_nodes(
+        &self,
+        source_file: &weldr::SourceFile,
+        transform: weldr::Mat4,
+        source_map: &weldr::SourceMap,
+        gltf: &mut gltf::Gltf,
+        buffer: &mut Vec<u8>,
+        mesh_cache: &mut HashMap<String, Option<usize>>,
+    ) -> u32 {
+        let matrix = if transform == weldr::Mat4::from_scale(1.0) {
+            None
+        } else {
+            Some([
+                transform.x.x,
+                transform.x.y,
+                transform.x.z,
+                transform.x.w,
+                transform.y.x,
+                transform.y.y,
+                transform.y.z,
+                transform.y.w,
+                transform.z.x,
+                transform.z.y,
+                transform.z.z,
+                transform.z.w,
+                transform.w.x,
+                transform.w.y,
+                transform.w.z,
+                transform.w.w,
+            ])
         };
-        offset += vertex_buffer.size;
-
-        let mut index_buffers = vec![];
-        if !geometry_cache.line_indices.is_empty() {
-            let buffer = GeometryBuffer {
-                size: line_indices_bytes.len(),
-                offset,
-                stride: 4,
-                component_type: gltf::ComponentType::UnsignedInt,
-                attribute_type: gltf::AttributeType::Scalar,
-            };
-            offset += buffer.size;
-            index_buffers.push(buffer);
-        }
-        if !geometry_cache.triangle_indices.is_empty() {
-            let buffer = GeometryBuffer {
-                size: triangle_indices_bytes.len(),
-                offset,
-                stride: 4,
-                component_type: gltf::ComponentType::UnsignedInt,
-                attribute_type: gltf::AttributeType::Scalar,
-            };
-            //offset += buffer.size;
-            index_buffers.push(buffer);
+
+        let node_index = gltf.nodes.len();
+        let node = gltf::Node {
+            name: Some(source_file.filename.clone()),
+            children: Vec::new(),
+            mesh_index: None,
+            matrix,
+        };
+        gltf.nodes.push(node);
+
+        // TODO: Check the part type rather than the extension.
```

Comment:
There's a special comment command for determining if the file is an official or unofficial "part". This should also be updated to check if each files contains geometry commands since apparently some files embed geometry inline for things like hoses and tubes. I'm not sure how trivial this would be to implement. Calling `root_file.iter(...)` can create enough geometry to crash many applications, so at least some level of instancing is necessary for medium to large models.

Reply:
There's more than one type that determines if it is a part. Ideally, we could just create a mesh for each ldraw file. In practice, this creates an unmanageable number of objects in the scene when importing. Stopping the recursion at the "part" level uses more memory but makes the file easier to work with. Whether the overhead matters depends on the application.